### PR TITLE
Add option to execute commands inside a container

### PR DESCRIPTION
**Pull request by querti**
Also, update the documentation reflecting the new changes, define
a Dockerfile of image capable of running all necessary skopeo
operations, and start using executors as context managers
(to perform self-cleanup).

### CODE REVIEW

File: pubtools/_quay/command_executor.py
Comment (neithere) [inline pubtools/_quay/command_executor.py:288]: `s/tt/rt/`

also it could be useful to mention that `~/.docker/` is used by default. Maybe even make it the default arg value instead of `None` and simply `os.path.expanduser(cert_path)`?

File: pubtools/_quay/tag_docker.py
Comment (neithere) [inline pubtools/_quay/tag_docker.py:735]: Is there a chance for the `run()` method to include more conditions over time? In that case it may eventually become hard to understand even despite the pretty good documentation, so it would make sense to extract the loop body to separate method(s). Otherwise it's fine.

File: pubtools/_quay/tag_images.py
Comment (neithere) [inline pubtools/_quay/tag_images.py:284]: Isn't this a longer way to say `executor_class = LocalExecutor`? :)

File: pubtools/_quay/tag_images.py
Comment (neithere) [inline pubtools/_quay/tag_images.py:287]: Debug print?

File: docker/skopeo_runner/Dockerfile
Comment (rohanpm) [inline docker/skopeo_runner/Dockerfile:1]: Is this the image you intend to use in production?  Can you share some info on how it'll be built?  It is best to have some kind of plan how to ensure this will be regularly rebuilt, and updated when F33 isn't supported any more.

In another repo I've had success with approach:

- use a specific digest in FROM
- activate renovatebot
- it will automatically bump the digest when a new version is available, e.g. https://github.com/rohanpm/ssh-prometheus-exporter/pull/9
- github updates trigger quay rebuild at https://quay.io/repository/rmcgover/ssh-prometheus-exporter

That way the repo keeps track of a known good image, and there's some automated prompt to get it regularly rebuilt.

File: pubtools/_quay/command_executor.py
Comment (rohanpm) [inline pubtools/_quay/command_executor.py:309]: What kind of values are you intending to provide as the image ?

If you want the possibility of images being updated after first use, should there be some 'pull' operation here?  On the other hand, the convention could be that you just have to update image in target settings if you ever want it to be updated.  But I have some doubts whether anyone would regularly do that.

File: docker/skopeo_runner/Dockerfile
Comment (zanssa) [inline docker/skopeo_runner/Dockerfile:3]: It's a metadata about the image. So, we could start with adding maintainer & description labels. 

LABEL Maintainer=" " \
           Description=" "

File: docker/skopeo_runner/Dockerfile
Comment (zanssa) [inline docker/skopeo_runner/Dockerfile:7]: in docker images, I prefer to do the installation & clean up in one layer.

Comment (zanssa) [inline docker/skopeo_runner/Dockerfile:7]:
```suggestion
RUN dnf install -y skopeo && dnf -y clean all
```

File: docker/skopeo_runner/Dockerfile
Comment (zanssa) [reply docker/skopeo_runner/Dockerfile:1]: you could also use github action to start building the image & pushing it to quay whenever there is a change in your dockerfile. here is an example: https://github.com/zanssa/iib-deployment/blob/main/.github/workflows/quay.yml

**Comment by twaugh** [comment]
Naively asking: is there a reason `docker` is being used instead of `podman`?

File: pubtools/_quay/command_executor.py
Comment (midnightercz) [inline pubtools/_quay/command_executor.py:405]: did you consider using docker ENV ?

**Review by midnightercz — CHANGES_REQUESTED** [review]

File: pubtools/_quay/command_executor.py
Comment (querti) [reply pubtools/_quay/command_executor.py]: Fixed typo and added information about default location. This part is mostly copy-pasted from pub so I'd prefer to leave it as is, if that's okay.

File: pubtools/_quay/tag_docker.py
Comment (querti) [reply pubtools/_quay/tag_docker.py:735]: I don't expect this logic to become more complex over time, but you never know. I'd say let's leave the refactoring out of this PR and do it once it becomes necessary. What do you think?

File: pubtools/_quay/tag_images.py
Comment (querti) [reply pubtools/_quay/tag_images.py:284]: :D . But I'd prefer to keep it this way because maybe LocalExecutor will accept some constructor arguments in the future. Also, I feel the code is a bit less convoluted when all executors are created the same way. Is it okay?

File: pubtools/_quay/tag_images.py
Comment (querti) [reply pubtools/_quay/tag_images.py]: Yep, removed.

File: docker/skopeo_runner/Dockerfile
Comment (querti) [reply docker/skopeo_runner/Dockerfile]: I added renovatebot settings and Github action for building it (just as a test). I'll add a Quay trigger after this PR is merged.

File: pubtools/_quay/command_executor.py
Comment (querti) [reply pubtools/_quay/command_executor.py:311]: I added a pull operation. Testing indicates that it's not a very expensive operation if the image is already downloaded.

File: docker/skopeo_runner/Dockerfile
Comment (querti) [reply docker/skopeo_runner/Dockerfile]: Added

Comment (querti) [reply docker/skopeo_runner/Dockerfile]: Added

File: pubtools/_quay/command_executor.py
Comment (querti) [reply pubtools/_quay/command_executor.py:312]: But will it be triggered if the executor is not used as context manager?

File: pubtools/_quay/command_executor.py
Comment (querti) [reply pubtools/_quay/command_executor.py:407]: I suppose it would be a viable solution as well, though I would prefer to keep the current one, simply because it's already implemented and tested. Is that ok?

File: pubtools/_quay/command_executor.py
Comment (midnightercz) [reply pubtools/_quay/command_executor.py:312]: It won't but you have stuff in __exit__ anyway, so you need to use it as context manager

File: pubtools/_quay/command_executor.py
Comment (midnightercz) [reply pubtools/_quay/command_executor.py:407]: sure, I was just wondering if was any problem with ENV or anything which prevented you from using it

**Review by midnightercz — APPROVED** [review]

**Comment by querti** [comment]
> Naively asking: is there a reason `docker` is being used instead of `podman`?

First reason is that pub-docker-daemon uses docker, not podman. Second reason is that Docker SDK for Python is used (https://docker-py.readthedocs.io/en/stable/), and I'm not sure if it's actually compatible with podman. It very well might be though.

File: pubtools/_quay/command_executor.py
Comment (querti) [reply pubtools/_quay/command_executor.py:312]: Well, exit only performs a cleanup, so it can be used as a regular class if the user doesn't mind zombie containers being left behind.

File: pubtools/_quay/command_executor.py
Comment (emilyzheng) [inline pubtools/_quay/command_executor.py:338]: exec_api installed by python-docker-py-1.7.2-1.el6eng.noarch doesn't have argument 'demux'.

File: pubtools/_quay/command_executor.py
Comment (querti) [reply pubtools/_quay/command_executor.py]: Added the correct import based on the version of docker.

File: pubtools/_quay/command_executor.py
Comment (querti) [reply pubtools/_quay/command_executor.py:300]: I checked target settings and it seems that specific version is not actually set, so "auto" is always used. I added an argument to always use "auto".

File: pubtools/_quay/command_executor.py
Comment (querti) [reply pubtools/_quay/command_executor.py]: Good point, I removed the parameter. The unfortunate side effect is that now we can't differentiate stdout and stderr. To maintain consistency with other classes, I duplicated the output to be returned as tuple. I think this change shouldn't affect any of the current use-cases.

File: pubtools/_quay/command_executor.py
Comment (neithere) [inline pubtools/_quay/command_executor.py:20]: Not sure if inheritance is better than plain assignment here.

File: pubtools/_quay/command_executor.py
Comment (neithere) [inline pubtools/_quay/command_executor.py:331]: s/suppoer/support/

File: pubtools/_quay/command_executor.py
Comment (neithere) [inline pubtools/_quay/command_executor.py:352]: Shouldn't the result depend on the version?

**Review by neithere — APPROVED** [review]

File: pubtools/_quay/command_executor.py
Comment (querti) [reply pubtools/_quay/command_executor.py]: fixed

File: pubtools/_quay/command_executor.py
Comment (querti) [reply pubtools/_quay/command_executor.py:352]: The main issue is that older versions of the Python dependency didn't even support this option.

**Pull request merged**
